test/e2e/backup_restore_test.go

```diff
@@ -0,0 +1,232 @@
+package e2e
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/iancoleman/strcase"
+	v1 "github.com/infinispan/infinispan-operator/pkg/apis/infinispan/v1"
+	v2 "github.com/infinispan/infinispan-operator/pkg/apis/infinispan/v2alpha1"
+	cconsts "github.com/infinispan/infinispan-operator/pkg/controller/constants"
+	ispn "github.com/infinispan/infinispan-operator/pkg/infinispan"
+	ispnclient "github.com/infinispan/infinispan-operator/pkg/infinispan/client/http"
+	tconst "github.com/infinispan/infinispan-operator/test/e2e/constants"
+	tutils "github.com/infinispan/infinispan-operator/test/e2e/utils"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type clusterSpec func(name, namespace string, clusterSize int) *v1.Infinispan
+
+func TestBackupRestore(t *testing.T) {
+	t.Run(string(v1.ServiceTypeDataGrid), testBackupRestore(datagridService))
+	t.Run(string(v1.ServiceTypeCache), testBackupRestore(cacheService))
+}
+
+func testBackupRestore(clusterSpec clusterSpec) func(*testing.T) {
+	return func(t *testing.T) {
+		t.Parallel()
+		// Create a resource without passing any config
+		name := strcase.ToKebab(strings.Replace(t.Name(), "/", "", 1))
+		namespace := tconst.Namespace
+		clusterSize := 2
+		numEntries := 100
+
+		// 1. Create initial source cluster
+		sourceCluster := name + "-source"
+		infinispan := clusterSpec(sourceCluster, namespace, clusterSize)
+		testKube.Create(infinispan)
+		defer testKube.DeleteInfinispan(infinispan, tconst.SinglePodTimeout)
+		waitForPodsOrFail(infinispan, clusterSize)
+
+		// 2. Populate the cluster with some data to backup
+		protocol := getSchemaForRest(infinispan)
+		cluster := newCluster(cconsts.DefaultOperatorUser, infinispan.GetSecretName(), protocol, testKube.Kubernetes)
+		cacheName := "someCache"
+		populateCache(cacheName, sourceCluster+"-0", numEntries, cluster.Client)
+
+		// 3. Backup the cluster's content
+		backupSpec := &v2.Backup{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+			Spec: v2.BackupSpec{
+				Cluster: sourceCluster,
+			},
+		}
+		testKube.Create(backupSpec)
+		defer testKube.DeleteBackup(backupSpec)
+
+		waitForValidBackupPhase(name, namespace, v2.BackupInitializing)
```

Comment:
@ryanemerson I'm not sure that all phase waits will work correctly in faster cluster test. The Backup/Restore `Status.Phase` implemented as state-machine and could be a situation while we are waiting for one state, but operator already on another state. We can schedule meeting to discuss this issue if we can't fit it into F2F time.